### ISSUE TITLE
Make control-c to preserve the contents of the line and start new line

### ISSUE THREAD

**Issue by obfuscated** [fish-shell/fish-shell#2904]
In bash if you type something and press ctrl-c then the content of the line is preserved and the cursor is moved to a new line. In fish the ctrl-c just clears the line. For me the behaviour of bash is a bit better, because it allows me to type something then press ctrl-c and I have the typed string in the log for further reference.

It looks something like:
>$ cd
>$ ls -l ^C
>$ ls -la
 
Would it be possible to add this feature in fish, I'm fine if this is a non default configuration option.

**Comment by ridiculousfish** [comment]
Makes sense to me

**Comment by krader1961** [comment]
It's a key binding:

```
bind \cc 'commandline ""'
```

The obvious solution is to change the binding to insert a "#" at the start of the line to turn it into a comment then enter the command. Unfortunately, I can't find a way to make the following work:

```
bind \cC "commandline -f beginning-of-line; commandline -i '#'; commandline -f end-of-line; commandline -f \r"
```
The insertion occurs before moving to the beginning of the line. It looks like some enhancements will be needed to make this work. I'm going to leave this labeled as a question for a couple of days to see if someone like @faho knows how to make this work using the current fish code.

**Comment by floam** [comment]
I found this somewhere one time and saved it. 

> bind \cc 'echo; commandline | cat; commandline ""; commandline -f repaint'

**Comment by faho** [comment]
>The obvious solution is to change the binding to insert a "#" at the start of the line to turn it into a comment

He... I've been playing around with this just yesterday, and what I came up with is

```fish
function switch-comment-commandline --description 'Comment/Uncomment the current or every line'
	set -l cmdlines (commandline)
	set -l cmdlines2 (commandline -c)
	if test "$cmdlines" = "$cmdlines2"
		commandline -r (printf '%s\n' '#'$cmdlines | string replace -r '^##' '')
	else
		set -l linenum (count $cmdlines2)
		set cmdlines[$linenum] (string replace -r '^##' '' -- '#'$cmdlines[$linenum])
		commandline -r $cmdlines
	end
end
```

The heart of the matter being `commandline -r (printf '%s\n' '#'$cmdlines | string replace -r '^##' '')`. The rest is pretty much all to support multi-line commandlines.

For this case, all you really need is something like `commandline -r '#'(commandline)` and an additional `commandline -f execute` for the binding.

**Comment by krader1961** [comment]
Thanks, @faho and @floam. I'll play with those solutions today and let you know if I find any problems. This capability is something I got into the habit of using long ago when I switched to ksh and then zsh and missed having in fish.

I'm also going to give some thought to how we can make these types of user contributions easier to discover. In the recent past I've seen several examples of this sort that don't warrant becoming default behavior but should be documented (outside of github issues) so people can more readily find them. Perhaps we need something like a "share/user_contrib" directory and someplace in the documentation to mention them.

**Comment by krader1961** [comment]
The only solution that works (mostly) as advertised for me is `bind \cc 'echo; commandline ""'`. I say "mostly" because if you enter

```
if true
    echo yes
```

then press [ctrl-C] you're left with just the first line visible. @floams's second binding doesn't put anything into my command history. That's the point of the "turn the partial input into a comment" approach. If you do that then the comment gets inserted into the command history which makes it retrievable. But I couldn't get @faho's function to work in vi-mode.

**Comment by jakwings** [comment]
An improved echo:

```fish
bind \cc 'commandline -a ""; for i in (seq (commandline -L)); echo; end; commandline ""'
```

**Comment by faho** [comment]
>But I couldn't get @faho's function to work in vi-mode.

Yeah, I've got some weird issues there as well. Will investigate. Though one binding that works for me is `bind s 'switch-comment-commandline' -m default`. If I change that to "\cs" it doesn't.

**Comment by faho** [comment]
@krader1961: I've managed to get it to work - I always confuse "-m" and "-M".

Can you try `bind -M default \cs 'switch-comment-commandline'`?

**Comment by krader1961** [comment]
@faho: The problem is that it's always taking the "else" branch and that code only affects the line with the cursor. So it works for a single line but not multiple lines. Not sure what your intent was but when I used this feature in ksh93 it always comments the entire command. So I just simplified your function to the following and get the behavior I expect:

```
set -l cmdlines (commandline)
commandline -r (printf '%s\n' '#'$cmdlines | string replace -r '^##' '')
commandline -f execute
```

**Comment by krader1961** [comment]
@jakwings: That's the ticket. The only thing I changed was adding "^C" to make the output resemble what you get from bash which makes it clearer the command was canceled:

```
function cancel-commandline
    commandline -C 2147483647
    for i in (seq (commandline -L))
        echo '^C'
    end
    commandline ""
end
```

That definitely seems like something we should make the fish default. Whereas @faho's function is probably better left as a "user contributed, unsupported" function.

**Comment by floam** [comment]
That's great! Its better than the default. OBJECTIVELY!

**Comment by floam** [comment]
Wow, I love that!

I am going with this, which displays ^C how I like and shows it only once if I hit it while editing a multiline prompt (like editing this function), instead of on the line I was editing and the beginning of every line thereafter. .

**Comment by Darkshadow2** [comment]
@krader1961 That _almost_ works. If you have a single line command that is long enough to wrap, it only echoes the first "line" and the rest is lost (and the ^C is not shown).

**Comment by krader1961** [comment]
Hmmm, yes, but the line doesn't have to literally wrap. It's enough for it to be longer than the space remaining after the prompt thus forcing fish to move the command to a line below the prompt. The best I can do is add an extra echo:

```
function cancel-commandline
    commandline -C 2147483647
    for i in (seq (commandline -L))
        echo '^C'
    end
    echo ''
    commandline ""
end
```

There doesn't seem to be a way to tell if fish has repositioned the text onto the line following the prompt. The above function adds an extra blank line when cancelling a single non-wrapped line. It produces the desired output for a wrapped (repositioned) line as well as if there are multiple lines.

**Comment by Darkshadow2** [comment]
Hmm, maybe this:

```fish
function cancel-commandline
   set -l len (string length -- (commandline))
   if test $len -eq 0; return; end
   commandline -C 2147483647
   for i in (seq (commandline -L))
      echo '^C'
   end
   while test (math $len - $COLUMNS) -gt 0
      echo
      set len (math $len - $COLUMNS)
   end
   commandline ""
end
```

That keeps it from having an empty line if short, but isn't perfect since it doesn't take into account the length of the prompt.

**Comment by faho** [comment]
>The problem is that it's always taking the "else" branch and that code only affects the line with the cursor. So it works for a single line but not multiple lines. Not sure what your intent was but when I used this feature in ksh93 it always comments the entire command.

I don't see that. The idea was that you'd want to comment or uncomment the entire buffer if the cursor is at the end, or just the line you are at if it is not.

`commandline` and  `commandline -c` should be equal if there's nothing after the cursor to cut, which should mean that the cursor is at the very end. And that's just how it works for me.

**Comment by ridiculousfish** [comment]
This is a nice fix! But the bright yellow is quite striking. I would like to make it gray instead. Any agreement or objections?

**Comment by krader1961** [comment]
The yellow was intentional since in western cultures it is associated with concepts like "warning" and "caution" and the strong visibility of the bright yellow makes it really obvious the pending command was cancelled. On the other hand until we can make `set_color` behave more sensibly with respect to eight color terminals `grey` is a safer choice.

**Comment by floam** [comment]
I like the yellow, it never struck me as too icky. Maybe it's just my colors?
<img width="571" alt="screenshot 2016-04-29 at 2 02 36 pm" src="https://cloud.githubusercontent.com/assets/291142/14929610/35a863c4-0e13-11e6-9b03-454c4e6cebb7.png">

I don't mind grey - my version of this often looked grey. I had done this:

<img width="251" alt="screenshot 2016-04-29 at 2 04 33 pm" src="https://cloud.githubusercontent.com/assets/291142/14929635/58e32a9a-0e13-11e6-9d87-0ee2aae9c58d.png">

`tput rev; echo "^C"; tput sgr0`

But I prefer actually inverting so that a user's terminal background color is unlikely to affect it. I sometimes use a grey background.


<img width="265" alt="screenshot 2016-04-29 at 2 04 24 pm" src="https://cloud.githubusercontent.com/assets/291142/14929638/59041af2-0e13-11e6-985f-4adad61dc29c.png">
<img width="247" alt="screenshot 2016-04-29 at 2 04 09 pm" src="https://cloud.githubusercontent.com/assets/291142/14929637/5903e29e-0e13-11e6-95f4-b01031739e19.png">


Is there a good reason why I shouldn't make `set_color` capable of this?

**Comment by gechr** [comment]
Why not just make it configurable, like other colours? Then the default doesn't really matter, as long as it's sensible.

Different people have different terminals, with varying capabilities & colour schemes. You can't cater to everyone.

**Comment by krader1961** [comment]
As @floam points out we really shouldn't be hardcoding the color in that script. We should be using one of the colors from the theme selected by the user. I didn't do that in my original commit because none of the existing theme color variables made sense semantically.

**Comment by floam** [comment]
Unfortunately a theme can't know what color someone's terminal uses as a foreground and background color. That's why I prefer `tput rev` -- it'd be nice if fish could do that where it works else fallback, through some kind of `set_color inverted` or something.

**Comment by krader1961** [comment]
> Unfortunately a theme can't know what color someone's terminal uses as a background color.

True, but presumably the user is using a theme that takes their terminal background color into account. But that returns us to my point that all of the existing theme color vars appear to be intended solely for fg use. In fact, I can't find any use of `set_color -b` in the standard fish scripts other than the one I added to *share/functions/__fish_cancel_commandline.fish*. Which surprises me.

So I'm inclined to agree with @floam that the only truly portable way to handle this that has the desired visual impact is to add something like a `--reverse` option to `set_color` or a pseudo-color named `reverse` and use that.

**Comment by krader1961** [comment]
I'll add a change to wrap the "^C" in `(tput smso)` and `(tput rmso)` and open an issue to track enhancing `set_color`. The cost of invoking an external program shouldn't be an issue in this case since it will be done infrequently and tput is pretty fast.

**Comment by ridiculousfish** [comment]
We don't need a big warning that the command was cancelled (zsh prints nothing at all here). It's fine to report it, but I find that the bright yellow draws too much of my attention and is distracting.

I like foreground gray (aka "brgrey"), which will match the newline indicator:

![screen shot 2016-04-29 at 2 59 33 pm](https://cloud.githubusercontent.com/assets/920838/14930722/fb599d2a-0e1a-11e6-9365-ea25aa4d3f55.png)

for reference, current behavior
![screen shot 2016-04-29 at 3 08 46 pm](https://cloud.githubusercontent.com/assets/920838/14930957/93fe6b36-0e1c-11e6-99ea-9d02e3cf917e.png)

I also like the idea of using Unicode here, perhaps ⦰. Though this will require some machinery to check whether Unicode is supported.
![screen shot 2016-04-29 at 3 10 44 pm](https://cloud.githubusercontent.com/assets/920838/14930952/8fba5210-0e1c-11e6-8beb-2f4eab51dda3.png)

As krader suggests, let's also make this a configurable color, which can cover both the newline and cancel signs. Maybe "fish_line_sigils"?

**Comment by krader1961** [comment]
Let's open a new issue to discuss enhancing themes to include support for background colors and special characters for special situations. Obviously we'll want a cheap and easy way for for functions to determine if they can safely use non-ascii chars. In the meantime, so we can get 2.3.0 out the door with this facility, the safest solution is the one @floam proposed that I mentioned in my previous comment: use `tput smso` to just invert fg/bg.

**Comment by floam** [comment]
I'm sorry, by zsh's behavior, I meant what it does with special characters in the line editor.
<img width="112" alt="screenshot 2016-04-29 at 3 20 18 pm" src="https://cloud.githubusercontent.com/assets/291142/14931144/fe6394f0-0e1d-11e6-8157-df228b6e6947.png">

**Comment by floam** [comment]
I do like the unicode idea. But I don't think it should be dim.

**Comment by ridiculousfish** [comment]
FYI this new control-C behavior is not in 2.3.0

**Comment by floam** [comment]
A dimmed newline indicator like we have makes sense to me - but I think something marking an action a user took should not have such a passive styling.

**Comment by krader1961** [comment]
> FYI this new control-C behavior is not in 2.3.0

You're right. I thought Zanchey included it when he cherry-picked the comment/uncomment change. Nonetheless, until we can reach a consensus on how to augment themes to support these types of use cases and so this works sensibly with eight color terminals like the Linux console VT we should just use tput to set standout mode for the ^C indicator.

**Comment by ridiculousfish** [comment]
Sure, let's try the standout mode and see how it feels.

**Comment by floam** [comment]
(FWIW, I still think I personally prefer yellow or red background (with white text, even) over what I get from `tput rev`/`tput smso` - but that'd be better for me to overwrite in a theme later. I think the color-agnostic tput thing is the better default though. (Also, I've said it before but I'll say it again - I think it would be nice if for that "dimmed" newline indicator thing, and autosuggestions, if fish effectively did `tput dim` instead of assuming a color unless a theme thinks it knows better. This is another thing `set_color` can't do.)

**Comment by krader1961** [comment]
I agree with you @floam. The nice thing about fish is it's trivial to put a copy of a standard function in your personal *~/config/fish/functions* directory and hack it to suit one's personal preferences so if I want bright yellow I can still have that while everyone else gets a reasonable experience :-)

**Comment by gechr** [comment]
> [...] a copy of a standard function in your personal ~/config/fish/functions directory [...]

@krader1961 While this works, I've personally been stung by doing it in the past. The problem is, you then stop receiving enhancements/fixes for that function and (in my case) completely forget you've taken a copy of the buggy default in the first place, which subsequently gets fixed upstream, and you don't receive the update.

In this case, would it not be better to add an empty variable in place of the colour (e.g. `$__fish_cancel_commandline_colour`), which can then be set in the user's `config.fish`? Then, all that needs to be set is the variable, rather than a copy of the whole function.

**Comment by floam** [comment]
> I'll add a change to wrap the "^C" in (tput smso) and (tput rmso) and open an issue to track enhancing set_color.

@krader1961 did you file that?

**Comment by cjxgm** [comment]
I hate the new behavior. How can I customize it back to the old "wipe out the current line" behavior?

**Comment by krader1961** [comment]
@cjxgm: Just add the following to your personal fish config:

```
function fish_user_key_bindings
    bind \cc 'commandline ""'
end
```

In hindsight this should have been added to the fish FAQ.

**Comment by rnhmjoj** [comment]
I added
```
function fish_user_key_bindings
    bind \cc 'commandline ""'
end
```
but there's must be something wrong: it behaves very strangely. Probably because I have a two-lines prompt. Sometimes ^C clears the line (old behavior), sometimes it starts a new line (new behavior) and sometimes it overwrites the last line with the current and start a new line.
